specifyweb/frontend/js_src/lib/components/WorkBench/WbSpreadsheet.tsx

```diff
@@ -0,0 +1,501 @@
+    readonly physicalCols: RA<number>;
+    readonly mappingPath: MappingPath;
+    readonly ids: RA<number>;
+    readonly key: string;
+  }>();
+  const [disambiguationPhysicalRow, setPhysicalRow] = React.useState<number>();
+  const [disambiguationResource, setResource] =
+    React.useState<Collection<AnySchema>>();
+  const [
+    noDisambiguationDialog,
+    openNoDisambiguationDialog,
+    closeNoDisambiguationDialog,
+  ] = useBooleanState();
+  const [disambiguationDialog, openDisambiguation, closeDisambiguation] =
+    useBooleanState();
+
+  const openDisambiguationDialog = () => {
+    if (mappings === undefined || !hotRef.current) return;
+
+    const [visualRow, visualCol] = getSelectedLast(
+      hotRef.current?.hotInstance!
+    );
+    const physicalRow = hotRef.current?.hotInstance!.toPhysicalRow(visualRow);
+    const physicalCol =
+      hotRef.current?.hotInstance!.toPhysicalColumn(visualCol);
+
+    const matches = validation?.uploadResults.ambiguousMatches[
+      physicalRow
+    ].find(({ physicalCols }) => physicalCols.includes(physicalCol));
+    if (matches === undefined) return;
+    const tableName = getTableFromMappingPath(
+      mappings.baseTable.name,
+      matches.mappingPath
+    );
+    const table = strictGetTable(tableName);
+    const resources = new table.LazyCollection({
+      filters: { id__in: matches.ids.join(',') },
+    }) as Collection<AnySchema>;
+
+    (hasTablePermission(table.name, 'read')
+      ? resources.fetch({ limit: 0 })
+      : Promise.resolve(resources)
+    ).then(({ models }) => {
+      if (models.length === 0) {
+        openNoDisambiguationDialog();
+        return;
+      }
+
+      // Re-enable this once live validation is available again:
+      /*
+       * Disable "Apply All" if validation is still in progress.
+       * This is because we don't know all matches until validation is done
+       */
+      /*
+       *Let applyAllAvailable = true;
+       *const applyAllButton = content.find('#applyAllButton');
+       *
+       *const updateIt = () => {
+       *  const newState = this.liveValidationStack.length === 0;
+       *  if (newState !== applyAllAvailable) {
+       *    applyAllAvailable = newState;
+       *    applyAllButton.disabled = !newState;
+       *    applyAllButton[newState ? 'removeAttribute' : 'setAttribute'](
+       *      'title',
+       *      wbText.applyAllUnavailable()
+       *    );
+       *  }
+       *};
+       *
+       *const interval = globalThis.setInterval(updateIt, 100);
+       * // onClose: globalThis.clearInterval(interval);
+       */
+
+      setDisambiguationMatches(matches);
+      setResource(resources);
+      setPhysicalRow(physicalRow);
+      openDisambiguation();
+    });
+  };
+
+  const contextMenuConfig = {
+    items: ensure<
+      IR<
+        | Handsontable.plugins.ContextMenu.MenuItemConfig
+        | Handsontable.plugins.ContextMenu.PredefinedMenuItemKey
+      >
+    >()(
```

Comment:
ensure() shouldn't be needed here.

```suggestion
const contextMenuConfig: DetailedSettings = {
  items:
```

the only catch is that you would have to replace `separator_1: '---------',` with `['separator_1' as 'redo']: '---------',` (and same for `separator_2`)

then you would also not need this type assertion down below anymore:

```diff
-    contextMenu={contextMenuConfig as Settings}
+    contextMenu={contextMenuConfig}
```

(import `DetailedSettings` from the same place you imported `Settings` from)

Reply:
`['separator_1' as 'undo']` throws an error `An object literal cannot have multiple properties with the same name.`:
```
undo: {
  disabled: () => hot?.isUndoAvailable() !== true || isReadOnly,
},
```

Reply:
so do `'separator_1' as 'redo'` in one place and `'separator_2' as 'undo'` in the other place?

Reply:
It throws an error for both the undo and redo keys
```
undo: {
   disabled: () => hot?.isUndoAvailable() !== true || isReadOnly,
},
redo: {
  disabled: () => hot?.isRedoAvailable() !== true || isReadOnly,
},
```

However, if I define the seperators as `['separator_1']: '---------'` and `['separator_2']: '---------',` everything works fine.

Reply:
From my understanding, the way [context menu is typed](https://handsontable.com/docs/javascript-data-grid/context-menu/#context-menu-with-specific-options) allows only one separator. For multiple separators either we suppress the typing error or do a type assertion like you did with an unused key eg: `'separator_1' as 'alignment'` instead of undo and redo

Reply:
> ['separator_1' as 'undo'] throws an error An object literal cannot have multiple properties with the same name.:

hmm, I don't see that error in typescript.
this type assertion is only for typescript - it should not affect runtime, and certainly not result in a runtime error

```ts
            ['separator_1' as 'separator']: '---------',
            fill_down: fillCellsContextMenuItem(hot!, 'down', isReadOnly),
            fill_up: fillCellsContextMenuItem(hot!, 'up', isReadOnly),
            ['separator_2' as 'separator']: '---------',
```

> From my understanding, the way [context menu is typed](https://handsontable.com/docs/javascript-data-grid/context-menu/#context-menu-with-specific-options) allows only one separator

indeed - it's a bit tricky in their case to type this case correctly so I don't blame them

> For multiple separators either we suppress the typing error or do a type assertion like you did with an unused key eg: 'separator_1' as 'alignment' instead of undo and redo

sounds good if that works